### ISSUE TITLE
Getting “Wait until the previous call completes” error while writing data to excel using office.js from Task pane application

### ISSUE THREAD

**Issue by kumarsai** [OfficeDev/office-js#483]
## Actual Issue:

After getting the latest updates from windows we are getting “Wait until the previous call completes” error for Office.js ( api method: Office.context.document.bindings.addFromNamedItemAsync - > getDataAsync).

Most of the users who are using this latest update are facing this issue.( for oder version it is working perfectly)

For the time being we removed windows update from USER machines and uninstalling and reinstalling the MS office.

## Just for info below are our observations:
We are seeing this when the name ranges to write are huge. This is happening only after the latest update. Before see, same code with the same name ranges number used to work fine.

Modified files are not reflecting on Task pane app for JS file ( but able to see html changes, not sure this a new update issue) even after doing below steps - Build project to generate JS files (from Type script). - Directly modified JS files. - Deleted IE browser cache. - Cleaned solution and rebuild again - Deleted temp files - System restarted.

Not able to attach debugger script process to visual studio (it is working fine with other system).

## Here is the version information where we are facing this issue:

https://i.stack.imgur.com/ks7Qp.jpg
https://i.stack.imgur.com/RHcDT.jpg

**Comment by tutieu** [comment]
Created a bug #3327208 to track this issue.

**Comment by kumarsai** [comment]
Friends is there any update on this issue??

**Comment by GauravSrivastava7** [comment]
Hi Office-JS team, there is new release of windows coming up on 10-May is office-js update bundled with it?

**Comment by sumurthy** [comment]
@kumarsai can you use the newer [API](https://docs.microsoft.com/en-us/javascript/api/excel/excel.bindingcollection?view=office-js#addfromnameditem-name--bindingtype--id-) and see if it works?  

Office-js is not bundled with Windows OS or Office application itself. It's available only on CDN and GitHub office-js package.

**Comment by kumarsai** [comment]
we pointed to new API [Link ](https://docs.microsoft.com/en-us/office/dev/add-ins/reference/requirement-sets/excel-api-requirement-sets) but issue is not resolved.

here is the combinations which we tested : 



comb | < Office Version -1904 | Office Version   -1904 (11601.20144)
-- | -- | --
**OS version   -1903** | Not   Working | Not   Working
**<= OS version - 1809** | Working | Working

is it due to Edge is used to load office addon?

Because this the major difference we observed between working and non working scenarios.

**Comment by kumarsai** [comment]
We found that windows 1903 build which uses Edge webView as host for task pane app has issues


but it seems to work fine when IE i used as host for addin


is there a way to swap Edge Webview to IE in 1903 windows build??

**Comment by kumarsai** [comment]
@sumurthy @tutieu  I created a sample application where I can Replicate the two issues, Click [here](https://microsoft-my.sharepoint.com/:u:/p/v-saarro/EdboDgV-nfZJgDZdofq-LD4BRrsdI7l0HSzPAnYlGRWvzg?e=xX5PDY) to download 

1. On reading multiple name ranges from excel getting error "Wait untile the previous call completes"
2. On writing multiple name ranges, write operations are get skipped for few name ranges but no Error.

Steps to run and replicate the issue:

1. Download zip file and extract it
2. Open solution in visual studio and run the application ( control + f5)
3. Open the excel file name : "Book 1.xlsm" from extracted folder, and click on show task pane.

On the Task pane app you will see 3 buttons { Write opp. Read opp, clear)
Click on "write opp" and check, it should write in all 32 name ranges

and then click on the "Read opp", you will get errors in text area.

**Comment by jargil** [comment]
@kumarsai  - I can confirm I can reproduce what you’re seeing, but after debugging the error you’re getting: “Wait until the previous call completes”, it literally means what is saying, you have to wait for the calls to complete. From debugging the difference is that the new Edge WebView control is processing the requests and pumping them to Excel much faster than IE was doing. Excel is exhausting its max number of concurrent calls before it can finish processing your requests.

I can confirm that the error is gone if I change the sample code to:

```javascript
function executeReadOpp() {
    //DemoObj.ReadExcelQPromise('testread2');
    //for (var i = 2; i < 68; i++) {

    //    DemoObj.ReadExcelQPromise('read' + i);
    //}
    var def = Q.defer();
    if (isCadTemplate) {
        ... code removed for legibility...
   }
    else {
       callReadExcelQPromise(2);   <-- This is the change. Your original code has a for loop that calls into DemoObj.ReadExcelQPromise directly.
    }
    $("#TraceArea").append('Fixed')
    return def.promise;
}

function callReadExcelQPromise(i) {
    if (i < 68) {
        DemoObj.ReadExcelQPromise('read' + i).then(function () {
            setTimeout(function () {
                callReadExcelQPromise(++i);
            }, 10);                    
        }());
    }
}
```

Ideally instead of the timeout, you’d pass a function to ReadExcelQPromise and call it once you resolve the promise. Basically the original for loop that you have instead of callReadExcelQPromise is storming Excel.

**Comment by kumarsai** [comment]
@jargil : Thanks for the solution, this issue is resolved.

But we are facing another issue  in new version of Office and OS.
 We are getting blank model-popup instead of browser tab when we click on any hyper links on Task pane.
https://microsoft-my.sharepoint.com/:u:/p/v-saarro/EdboDgV-nfZJgDZdofq-LD4BRrsdI7l0HSzPAnYlGRWvzg?e=hD7o08  (updated the same app)
Here is sample application we created to replicate this issue.
    _Run this application and click on Download button in Task pane._

This functionality is working in older version of OS/Excel

screen shot of download issue.
https://microsoft-my.sharepoint.com/:i:/p/v-saarro/EcVsZybQgG5Cty20vucCxQcBGA4tInFATxyJrDCmSkBZCg?e=4QsSAA

**Comment by jargil** [comment]
@kumarsai - to better track separate issues, could you please open a new GitHub issue describing what you're seeing about the popup? Feel free to assign to me once you have that. Please add as much information as you possibly can.
I'll go ahead and close this issue here. Thanks!
